### PR TITLE
Memory pair game

### PR DESCRIPTION
**Pull request by zalevskyi**
### Memory pair game

[Demo](https://zalevskyi.github.io/kottans-frontend/practice/memory-pair-game/index.html)
[Code base](https://github.com/zalevskyi/kottans-frontend/tree/main/practice/memory-pair-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    open() {this.div.classList.add('opened')}
+    close() {this.div.classList.remove('opened')}
+    hide() {this.div.classList.add('hidden')}
+    static match(card1, card2) {
+        if (card1.image_file === card2.image_file) return true
+        else return false
+    }
+}
+//Card deck for memory pair matching
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:22]: Redundant comment

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    shuffle() {
+        let currentDealImages = shuffleArray(this.imagesArray).slice(0, this.cardsQuantityInDeal / 2)
+        currentDealImages = shuffleArray(currentDealImages.concat(currentDealImages))
+        this.currentDealCards = currentDealImages.map((image, index) => new Card(image, index))
+        this.opened = []
+        this.matched_quantity = 0
+        this.flip_count = 0
+    }
+    placeOnBoard(board) {
+        //board is div element for appending child card elements
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:45]: Redundant comment

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            this.matched_quantity += this.opened.length
+            this._hideOpened()
+            if (this.matched_quantity == CARDS_QUANTITY) this.showResult(this.flip_count)
+        }
+    }
+    _closeOpened() {while(this.opened.length > 0) this.opened.pop().close()}
+    _hideOpened() {while(this.opened.length > 0) this.opened.pop().hide()}
+}
+function cardNumber(elemenId) {
+    //extracts Id number from HTML element id
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:67]: Redundant comment

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
+class Card {
+    constructor(image, index) {
+        //image - object with file name and alt text of the image
+        //index - card index in current deal of the deck
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:4]: Redundant comment

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+const CARDS_QUANTITY = 16
+const DECK = new Deck(IMAGES, CARDS_QUANTITY, showResult)
+
+BOARD.addEventListener('click', evt => {
+    if (evt.target.id==='start_game') {
+        evt.preventDefault()
+        startGame()
+    } else {
+        let cardId = cardNumber(evt.target.id)
+        if (cardId!==undefined) DECK.openCard(cardId)
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:109]:
```suggestion
        if (cardId) DECK.openCard(cardId)
```

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    openCard(cardId) {
+        this.flip_count++
+        if (this.opened.length==2) this._closeOpened()
+        this.currentDealCards[cardId].open()
+        this.opened.push(this.currentDealCards[cardId])
+        this._checkOpenedMatch()
+    }
+    _checkOpenedMatch() {
+        if (this.opened.length==2 && Card.match(this.opened[0], this.opened[1])) {
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:57]: will be more readable with destructuring
`const [firstCard, secondCard] = this.opened;`

use strict equality operator

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            this.matched_quantity += this.opened.length
+            this._hideOpened()
+            if (this.matched_quantity == CARDS_QUANTITY) this.showResult(this.flip_count)
+        }
+    }
+    _closeOpened() {while(this.opened.length > 0) this.opened.pop().close()}
+    _hideOpened() {while(this.opened.length > 0) this.opened.pop().hide()}
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:64]: format your code please, not readable

File: submissions/zalevskyi/memory-pair-game/app.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        this._checkOpenedMatch()
+    }
+    _checkOpenedMatch() {
+        if (this.opened.length==2 && Card.match(this.opened[0], this.opened[1])) {
+            this.matched_quantity += this.opened.length
+            this._hideOpened()
+            if (this.matched_quantity == CARDS_QUANTITY) this.showResult(this.flip_count)
+        }
+    }
+    _closeOpened() {while(this.opened.length > 0) this.opened.pop().close()}
```
Comment (babayK0) [inline submissions/zalevskyi/memory-pair-game/app.js:63]:
```suggestion
    _closeOpened() {while(this.opened.length) this.opened.pop().close()}
```

**Review by babayK0 — CHANGES_REQUESTED** [review]
Good job !
Watch my comments below)

**Comment by zalevskyi** [comment]
Thank you for the review!
All recommendations are implemented.
Please review once more

**Review by babayK0 — APPROVED** [review]
Good job!

**Pull request merged**